src/toxcore/tcp/packet.rs

```diff
@@ -473,6 +475,49 @@ impl ToBytes for OobReceive {
     }
 }
 
+/** Sent by client or relay server to server or further relay server.
+Same format as initial UDP onion packet 0x80 (from chapter 11), but TCP packet id is 0x08 instead.
+
+Packet type [`Kind::OnionRequest`](./enum.Kind.html).
+
+
+Serialized form:
+
+Length   | Content
+-------- | ------
+`1`      | 0x08
+`32`     | Nonce
+`32`     | Our Temporary DHT Key
```

Comment:
```
        case TCP_PACKET_ONION_REQUEST: {
            if (TCP_server->onion) {
                if (length <= 1 + crypto_box_NONCEBYTES + ONION_SEND_BASE * 2)
                    return -1;

                IP_Port source;
                source.port = 0;  // dummy initialise
                source.ip.family = TCP_ONION_FAMILY;
                source.ip.ip6.uint32[0] = con_id;
                source.ip.ip6.uint32[1] = 0;
                source.ip.ip6.uint64[1] = con->identifier;
                onion_send_1(TCP_server->onion, data + 1 + crypto_box_NONCEBYTES, length - (1 + crypto_box_NONCEBYTES), source,
                             data + 1);
            }

            return 0;
        }
```

`int onion_send_1( const Onion* onion, const uint8_t* plain, uint16_t len, IP_Port source, const uint8_t* nonce )`

`plain` starts from `data + 1 + crypto_box_NONCEBYTES`

```
int onion_send_1(const Onion *onion, const uint8_t *plain, uint16_t len, IP_Port source, const uint8_t *nonce)
{
    if (len > ONION_MAX_PACKET_SIZE + SIZE_IPPORT - (1 + crypto_box_NONCEBYTES + ONION_RETURN_1))
        return 1;

    if (len <= SIZE_IPPORT + SEND_BASE * 2)
        return 1;

    IP_Port send_to;

    if (ipport_unpack(&send_to, plain, len, 0) == -1)
        return 1;
...
...
```

So packet is:
```
1    0x08
32  Nonce
19  IP_Port
... rest ?
```

please check it

Reply:
TCP relays can only be node A or the first peer in the chain as the TCP relay is essentially acting as a gateway to the network. The data sent to the TCP Client module to be sent as a TCP onion packet by the module is different from the one sent directly via UDP. This is because it doesn't need to be encrypted (the connection to the TCP relay server is already encrypted).

Reply:
@fare or shall we leave IP_Port unparsed inside data?